tests/test_lowlevel.py

```diff
@@ -204,3 +204,39 @@ def test_use_proxy_from_environment(httpbin, var, scheme):
 
         # it had actual content (not checking for SOCKS protocol for now)
         assert len(fake_proxy.handler_results[0]) > 0
+
+
+def test_redirect_rfc1808_to_non_ascii_location():
+    path = u'š'
+    expected_path = quote(path.encode('utf8')).encode('ascii')
+    expected_path_py3 = b'%C3%85%C2%A1'
+
+    def redirect_resp_handler(sock):
+        consume_socket_content(sock, timeout=0.5)
+        location = u'//{0}:{1}/{2}'.format(host, port, path)
+        sock.send(
+            b'HTTP/1.1 301 Moved Permanently\r\n'
+            b'Content-Length: 0\r\n'
+            b'Location: ' + location.encode('utf8') + b'\r\n'
+            b'\r\n'
+        )
+        second_request = consume_socket_content(sock, timeout=0.5)
+
+        # currently Python3 not handling non-ASCII redirects (issue #3888)
+        if is_py3:
+            assert second_request.startswith(b'GET /' + expected_path_py3 + b' HTTP/1.1')
+        else:
+            assert second_request.startswith(b'GET /' + expected_path + b' HTTP/1.1')
```

Comment:
Ok, one last note. It's slightly nicer to save this into a mutable object in the enclosing scope (e.g. a list) and then put the assertion in the enclosing function rather than the socket handler. assertions in the socket handler *do* work, but it's usually nicer to let the socket handler run to completion rather than blow it up and possibly leak it.